pyccel/ast/headers.py

```diff
@@ -211,7 +230,27 @@ def build_argument(var_name, dc):
                 dtypes += [[i]]
             else:
                 raise TypeError('element must be of type UnionType or dict')
-        for args_ in product(*dtypes):
+
+        #TODO: handel the case of functions arguments
+        templates_names = [i.name for i in templates]
+        arg_codes = []
+        for iterx in product(*dtypes):
+            iterx = list(iterx)
+            old_values = {}
+            for i, j in enumerate(iterx):
+                if 'datatype' in j and j['datatype'] in templates_names and not j['datatype'] in old_values:
+                    tmplt = templates_names.index(j['datatype'])
+                    iterx[i] = templates[tmplt].args
+                    old_values[j['datatype']] = i
+                else:
+                    iterx[i] = [j]
+            for itery in product(*iterx):
+                itery = list(itery)
+                for j, i in enumerate(itery):
+                    if 'datatype' in i and i['datatype'] in templates_names:
+                        itery[j] = itery[old_values[i['datatype']]].copy()
+                arg_codes.append(itery)
```

Comment:
Can you add some comments to this section to explain what each step is doing please